### PR TITLE
Added blacklisted_paths option

### PR DESCRIPTION
**Pull request by schroedingerskatze42**
- adds a list of strings (regular expressions) to describe patterns of paths not to be cached.

| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    |yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | -
| Documentation   | -
| License         | MIT

#### What's in this PR?

Added an option 'blacklisted_paths'; a list of regular expressions to describe paths not to be cached. 

#### Why?

This is necessary if a third party resource is requested of providers, who don't get the usage of response cache headers. (But will return no-cache every time, for every endpoint)  

#### Example Usage

``` php
$cachePlugin = new CachePlugin($pool, $streamFactory, ['blacklisted_paths' => [
  '^\/this-path-shall\/not-been-cached\/.*',
]]);
```

#### Checklist

- [X] Updated CHANGELOG.md to describe BC breaks / deprecations | new feature | bugfix
- [X] Documentation pull request created: https://github.com/php-http/documentation/pull/274

### CODE REVIEW

File: src/CachePlugin.php [hunk 1]
```diff
@@ -72,10 +73,7 @@ public function __construct(CacheItemPoolInterface $pool, StreamFactory $streamF
         $this->streamFactory = $streamFactory;
 
         if (isset($config['respect_cache_headers']) && isset($config['respect_response_cache_directives'])) {
-            throw new \InvalidArgumentException(
-                'You can\'t provide config option "respect_cache_headers" and "respect_response_cache_directives". '.
-                'Use "respect_response_cache_directives" instead.'
-            );
+            throw new \InvalidArgumentException('You can\'t provide config option "respect_cache_headers" and "respect_response_cache_directives". '.'Use "respect_response_cache_directives" instead.');
```
Comment (dbu) [inline src/CachePlugin.php]: did the codestyle checker complain about this before? otherwise lets not touch it.

if codestyle complains, lets remove the concatenation and make it all one string.

File: src/CachePlugin.php [hunk 2]
```diff
@@ -353,13 +358,15 @@ private function configureOptions(OptionsResolver $resolver)
             'respect_response_cache_directives' => ['no-cache', 'private', 'max-age', 'no-store'],
             'cache_key_generator' => null,
             'cache_listeners' => [],
+            'blacklisted_paths' => [],  // restricted for
```
Comment (dbu) [inline src/CachePlugin.php]: please remove the comment, the phpdoc on the constructor should document this (plus the documentation change)

File: src/CachePlugin.php [hunk 3]
```diff
@@ -246,16 +244,23 @@ private function calculateResponseExpiresAt($maxAge)
     /**
      * Verify that we can cache this response.
      *
+     * @param RequestInterface  $request
      * @param ResponseInterface $response
      *
      * @return bool
      */
-    protected function isCacheable(ResponseInterface $response)
+    protected function isCacheable(RequestInterface $request, ResponseInterface $response)
```
Comment (dbu) [inline src/CachePlugin.php]: i wonder why this method is protected. extending the plugin sounds like a bad idea, but maybe that was the intention. as it is protected, this is technically a BC break. we could separate this into another method `isCacheableRequest` to avoid the BC break.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,11 @@
 # Change Log
 
+## 1.6.1 - 2019-11-25
```
Comment (dbu) [inline CHANGELOG.md]: this is a new feature and should lead to 1.7.0

please replace the date with "unreleased"

**Review by dbu — COMMENTED** [review]
thanks for contributing. this is rather an edge case, but as we do have modes of ignoring caching instructions from the server, it seems reasonable to me to have this config.

could you please add something to the test in spec/CachePluginSpec.php to have one example of blacklisting tested? and wdyt of splitting it into a separate method to avoid the BC break?

File: src/CachePlugin.php (hunk 1 above)
Comment (schroedingerskatze42) [reply src/CachePlugin.php]: yes, the code style checker complained about it; should already have been existed before. no magic button (auto format) used.

regarding the test: WIP, give me 15 minutes.

File: CHANGELOG.md (hunk 4 above)
Comment (schroedingerskatze42) [reply CHANGELOG.md]: done; also replaced 1.6.1 with 1.7.0; not quite sure, what is right here. feel free to enlighten me :-)

File: src/CachePlugin.php (hunk 3 above)
Comment (schroedingerskatze42) [reply src/CachePlugin.php]: you are right. i outsourced the checking of the request into the method isCacheableRequest and reset the "isCacheable" to it's original state.

The only thing, I am bothering about is the naming (one of the two hard things of computer science). If we add a method "isCacheableRequest", the method "isCacheable" should be renamed to "isCacheableResponse" and a new method "isCacheable" should take both arguments and return the result of isCacheableRequest && isCachableResponse().

This however would lead again to the BC, we would like to avoid.

File: src/CachePlugin.php (hunk 2 above)
Comment (schroedingerskatze42) [reply src/CachePlugin.php]: done

File: src/CachePlugin.php (hunk 1 above)
Comment (dbu) [reply src/CachePlugin.php]: the stupid thing is that it does a string concat on that one line now, for no good reason... can you remove the `'.'` in the middle please?

File: src/CachePlugin.php (hunk 3 above)
Comment (dbu) [reply src/CachePlugin.php]: yep. my idea would be that in version 2 we make the plugin final, this method private and can change the signature.

File: CHANGELOG.md (hunk 4 above)
Comment (dbu) [reply CHANGELOG.md]: that looks good, thanks.

File: src/CachePlugin.php [hunk 5]
```diff
@@ -266,6 +264,24 @@ protected function isCacheable(ResponseInterface $response)
...
     }
 
+    /**
+     * Verify that we can cache this request.
+     *
+     * @param RequestInterface  $request
+     *
+     * @return bool
+     */
+    protected function isCacheableRequest(RequestInterface $request)
```
Comment (dbu) [inline src/CachePlugin.php]: please make this method private.

**Comment by dbu** [comment]
apart from my nitpicks, i am happy with this.

@nobYsDarling can you please do the whitespace fix that codestyle is complaining about?

@Nyholm any inputs from your side or happy with this?

File: src/CachePlugin.php (hunk 1 above)
Comment (schroedingerskatze42) [reply src/CachePlugin.php]: done.

File: src/CachePlugin.php (hunk 5 above)
Comment (schroedingerskatze42) [reply src/CachePlugin.php]: done.

**Review by dbu — APPROVED** [review]
